lib/FindTests.js

```diff
@@ -0,0 +1,269 @@
+  // operates from the project root dir.
+  const globRelativeToProjectRoot = path.relative(
+    projectRootDir,
+    path.resolve(fileGlob)
+  );
+  return flatMap(
+    glob.sync(globRelativeToProjectRoot, {
+      cwd: projectRootDir,
+      nocase: true,
+      absolute: true,
+      ignore: ignoredDirsGlobs,
+      // Match directories as well and mark them with a trailing slash.
+      nodir: false,
+      mark: true,
+    }),
+    (filePath) =>
+      filePath.endsWith('/') ? findAllElmFilesInDir(filePath) : filePath
+  );
+}
+
+// Recursively search for *.elm files.
+function findAllElmFilesInDir(dir /*: string */) /*: Array<string> */ {
+  return glob.sync('**/*.elm', {
+    cwd: dir,
+    nocase: true,
+    absolute: true,
+    ignore: ignoredDirsGlobs,
+    nodir: true,
+  });
+}
+
+function findTests(
+  testFilePaths /*: Array<string> */,
+  project /*: typeof Project.Project */
+) /*: Promise<Array<{ moduleName: string, possiblyTests: Array<string> }>> */ {
+  return Promise.all(
+    testFilePaths.map((filePath) => {
+      const matchingSourceDirs = project.testsSourceDirs.filter((dir) =>
+        filePath.startsWith(`${dir}${path.sep}`)
+      );
+
+      // Tests must be in tests/ or in source-directories – otherwise they won’t
+      // compile. Elm won’t be able to find imports.
+      switch (matchingSourceDirs.length) {
+        case 0:
+          return Promise.reject(
+            Error(
+              missingSourceDirectoryError(
+                filePath,
+                project.elmJson.type === 'package'
+              )
+            )
+          );
+
+        case 1:
+          // Keep going.
+          break;
+
+        default:
+          // This shouldn’t be possible for package projects.
+          return Promise.reject(
+            new Error(
+              multipleSourceDirectoriesError(
+                filePath,
+                matchingSourceDirs,
+                project.testsDir
+              )
+            )
+          );
+      }
+
+      // By finding the module name from the file path we can import it even if
+      // the file is full of errors. Elm will then report what’s wrong.
+      const moduleNameParts = path
+        .relative(matchingSourceDirs[0], filePath)
+        .replace(/\.elm$/, '')
+        .split(path.sep);
+      const moduleName = moduleNameParts.join('.');
+
+      if (!moduleNameParts.every(Parser.isUpperName)) {
+        return Promise.reject(
+          new Error(
+            badModuleNameError(filePath, matchingSourceDirs[0], moduleName)
+          )
+        );
+      }
+
+      return Parser.extractExposedPossiblyTests(
+        filePath,
+        // We’re reading files asynchronously in a loop here, so it makes sense
+        // to use graceful-fs to avoid “too many open files” errors.
+        gracefulFs.createReadStream
+      ).then((possiblyTests) => ({
+        moduleName,
+        possiblyTests,
+      }));
+    })
+  );
+}
```

Comment:
Moved from Runner.js (really, Runner.js was renamed to FindTests.js because that was all it was doing). The only change is that the function takes a `project: Project` parameter instead of `sourceDirs: Array<string>, isPackageProject: boolean`.